lib/lf/uscalfe/lagr_fe.h

```diff
@@ -710,6 +710,455 @@ class FeLagrangePoint : public ScalarReferenceFiniteElement<SCALAR> {
+ *
+ * This is a specialization of ScalarReferenceFiniteElement.
+ * Refer to its documentation.
+ *
+ * The reference shape functions are combinations of the barycentric coordinate
+ * functions on the reference triangle.
+ *
+ * The first three shape functions are associated to the vertices,
+ * the other three shape functions to the edges of the triangle. There are
+ * no interior shape functions.
+ */
+template <typename SCALAR>
+class FeLagrangeO2Tria final : public ScalarReferenceFiniteElement<SCALAR> {
+ public:
+  FeLagrangeO2Tria(const FeLagrangeO2Tria&) = default;
+  FeLagrangeO2Tria(FeLagrangeO2Tria&&) noexcept = default;
+  FeLagrangeO2Tria& operator=(const FeLagrangeO2Tria&) = default;
+  FeLagrangeO2Tria& operator=(FeLagrangeO2Tria&&) noexcept = default;
+  FeLagrangeO2Tria() = default;
+  ~FeLagrangeO2Tria() override = default;
+
+  [[nodiscard]] base::RefEl RefEl() const override {
+    return base::RefEl::kTria();
+  }
+
+  [[nodiscard]] unsigned Degree() const override { return 2; }
+
+  /**
+   * @copydoc ScalarReferenceFiniteElement::NumRefShapeFunctions()
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions() const override { return 6; }
+
+  /**  @brief One shape function attached to each node
+   * and one to each edge of the triangle. There are no interior
+   * shape functions.
+   * @copydoc ScalarReferenceFiniteElement::NumRefShapeFunctions(dim_t)
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(dim_t codim) const override {
+    LF_ASSERT_MSG(codim <= 2, "Illegal codim" << codim);
+    return (codim == 0) ? 0 : 1;
+  }
+
+  /**  @brief One shape function attached to each node
+   * and one to each edge of the triangle. There are no interior
+   * shape functions.
+   * @copydoc
+   * ScalarReferenceFiniteElement::NumRefShapeFunctions(dim_t,sub_idx_t)
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(
+      dim_t codim, sub_idx_t /*subidx*/) const override {
+    LF_ASSERT_MSG(codim <= 2, "Illegal codim" << codim);
+    return (codim == 0) ? 0 : 1;
+  }
+
+  /** @copydoc ScalarReferenceFiniteElement::EvalReferenceShapeFunctions() */
+  [[nodiscard]] Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic>
+  EvalReferenceShapeFunctions(const Eigen::MatrixXd& refcoords) const override {
+    LF_ASSERT_MSG(refcoords.rows() == 2,
+                  "Reference coordinates must be 2-vectors");
+
+    const size_type n_pts(refcoords.cols());
+    Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic> result(6, n_pts);
+
+    Eigen::ArrayXd x0 = refcoords.row(0).array();
+    Eigen::ArrayXd x1 = refcoords.row(1).array();
+
+    result.row(0) = (2.0 * (1 - x0 - x1) * (0.5 - x0 - x1)).matrix();
+    result.row(1) = (2.0 * x0 * (x0 - 0.5)).matrix();
+    result.row(2) = (2.0 * x1 * (x1 - 0.5)).matrix();
+    result.row(3) = (4.0 * (1 - x0 - x1) * x0).matrix();
+    result.row(4) = (4.0 * x0 * x1).matrix();
+    result.row(5) = (4.0 * (1 - x0 - x1) * x1).matrix();
+
+    return result;
+  }
```

Comment:
I would put the implementation into the source file `lagr_fe.cc`, see also next comment.